_posts/2020-10-07-torch-modules/torch_modules.Rmd

```diff
@@ -0,0 +1,429 @@
+---
+title: "Using torch modules"
+description: > 
+ In this next-to-last installment of a mini-series introducing torch basics, we replace hand-coded matrix operations by modules, considerably simplifying our toy network's code. 
+author:
+  - name: Sigrid Keydana
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydanatorchmodules
+date: 10-07-2020
+categories:
+  - Torch
+  - R
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+editor_options: 
+  markdown: 
+    wrap: 72
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+[Last
+week](https://blogs.rstudio.com/ai/posts/2020-10-01-torch-network-from-scratch),
+we started learning about `torch` basics by coding a simple neural
+network from scratch, making use of just a single of `torch`'s features:
+*tensors*. [On
+Monday](https://blogs.rstudio.com/ai/posts/2020-10-05-torch-network-with-autograd),
+we immensely simplified the task, replacing manual backpropagation with
+*autograd*. Today, we *modularize* the network - in both the habitual
+and a very literal sense: Low-level matrix operations are swapped out
+for `torch` `module`s.
+
+## Modules
+
+From other frameworks (Keras, say), you may be used to distinguishing
+between *models* and *layers*. In `torch`, both are instances of
+`nn_Module()`, and thus, have some methods in common. For those thinking
+in terms of "models" and "layers", I'm artificially splitting up this
+section into two parts, for clarity of exposition. In reality though,
+there is no dichotomy: New modules may be composed of existing ones up
+to arbitrary levels of recursion.
+
+### Base modules ("layers")
+
+Instead of writing out an affine operation by hand -- `x$mm(w1) + b1`,
+say --, as we've been doing so far, we can create a linear module. The
+following snippet instantiates a linear layer that expects three-feature
+inputs and returns a single output per observation:
+
+```{r}
+library(torch)
+l <- nn_linear(3, 1)
+```
+
+The module has two parameters, "weight" and "bias". Both have already
+been initialized:
```

Comment:
Automatically by torch or is that what you did in the previous chunk?

Reply:
right, making that clearer...